### PR TITLE
Fixes #31629 - API for generating registration command

### PR DESCRIPTION
**Pull request by stejskalleos**
New `/api/registration_command` endpoint for generating Registration Command via CLI.

**How to test:**
```
curl -X POST foreman.example.com/api/registration_command \
     --user admin:changeme \
     -H 'Content-Type: application/json'
```

**Related PRs:**
* https://github.com/theforeman/hammer-cli-foreman/pull/558
* https://github.com/theforeman/foreman_remote_execution/pull/561/files
* https://github.com/Katello/katello/pull/9108

### CODE REVIEW

**Comment by stejskalleos** [comment]
[test katello]

**Review by chris1984 — DISMISSED** [review]
ACK, tested with UI and was able to register a host and also with curl:

```bash
curl -X POST https://devel.toledo.satellite.lab.eng.rdu2.redhat.com/api/registration_command --user admin:changeme -H 'Content-Type: application/json'

[vagrant@devel ~]$ curl -X POST https://devel.toledo.satellite.lab.eng.rdu2.redhat.com/api/registration_command --user admin:changeme -H 'Content-Type: application/json'
{"registration_command":"curl  -X GET 'https://devel.toledo.satellite.lab.eng.rdu2.redhat.com/register' -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJpYXQiOjE2MTIxOTI5ODQsImp0aSI6Ijg4MDM4NDA1NzFmZTU5MWY0ZjIzMTRjZWVjOWZiN2I0OWM3ODc5OTNkNzQxY2VjMTVlYWJjNjcxM2VlMWE2NGQiLCJleHAiOjE2MTIyMDczODQsInNjb3BlIjoicmVnaXN0cmF0aW9uI2dsb2JhbCByZWdpc3RyYXRpb24jaG9zdCJ9.ZhWpe516naVMjgUlVxPhqUsCKhJyvHpIA57WH-NCI_c' | bash"}
```

**Comment by chris1984** [comment]
[test katello]

**Comment by chris1984** [comment]
[test katello]

**Comment by chris1984** [comment]
@stejskalleos katello won't pass until this goes in, but acked and it works

File: app/controllers/api/v2/registration_command_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+module Api
+  module V2
+    class RegistrationCommandController < V2::BaseController
+      include Api::Version2
+      include Foreman::Controller::RegistrationCommand
+
+      api :POST, "/registration_command", N_("Generate global registration command")
+      param :registration_command, Hash, required: false, action_aware: true do
+        param :organization_id, :number, desc: N_("ID of the Organization to register the host in.")
```
Comment (ofedoren) [inline app/controllers/api/v2/registration_command_controller.rb]: Descriptions must not end with a dot.

File: app/controllers/concerns/foreman/controller/registration_command.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  def registration_args
+    args = registration_params.except(*ignored_query_args)
+
+    args['setup_insights'] = setup_insights.to_s
+    args['setup_remote_execution'] = setup_remote_execution.to_s
+
+    args.delete_if { |_, v| v.blank? }
+        .permit!
+  end
```
Comment (ofedoren) [inline app/controllers/concerns/foreman/controller/registration_command.rb]: I don't think this is needed... the same can be achieved within `registration_params` method if all you need is to remove empty/falsy params. Then you can use e.g. `registration_params.to_query`.

BTW, `[false, nil, "", "    ", [], {}].map(&:blank?) #=> [true, true. true, true, true. true]`.

UPD: Sorry, I see now that `setup_insights` return value is used... but still I question why that method exists...

File: app/controllers/concerns/foreman/controller/registration_command.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    organization = Organization.authorized(:view_organizations).find(registration_params['organization_id']) if registration_params['organization_id'].present?
+    location = Location.authorized(:view_locations).find(registration_params['location_id']) if registration_params['location_id'].present?
+    host_group = Hostgroup.authorized(:view_hostgroups).find(registration_params['hostgroup_id']) if registration_params["hostgroup_id"].present?
+    operatingsystem = Operatingsystem.authorized(:view_operatingsystems).find(registration_params['operatingsystem_id']) if registration_params["operatingsystem_id"].present?
+
+    Host.new(organization: organization, location: location, hostgroup: host_group, operatingsystem: operatingsystem).params
+  end
+
+  def setup_insights
+    return if registration_params['setup_insights'].to_s.blank?
```
Comment (ofedoren) [inline app/controllers/concerns/foreman/controller/registration_command.rb]: Just curiosity: why `.to_s` cast is needed?

```ruby
registration_params['setup_insights'] = nil
registration_params['setup_insights'].blank? #=> true
registration_params['setup_insights'] = false
registration_params['setup_insights'].blank? #=> true
# but
registration_params['setup_insights'] = false.to_s
registration_params['setup_insights'].blank? #=> false
```

Just a bit hard to follow why exactly in this way and what exactly you want to achieve.

File: app/controllers/concerns/foreman/controller/registration_command.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    from_host = host_config_params['host_registration_insights']
+    from_request = ActiveRecord::Type::Boolean.new.deserialize(registration_params['setup_insights'])
+
+    from_request if (from_request != from_host)
```
Comment (ofedoren) [inline app/controllers/concerns/foreman/controller/registration_command.rb]: Just to understand it right: this method returns one of 3 values: true, false, nil. If two of them mean the same (nil is considered as false, not a third different meaning) then why are here extra steps?... Moreover, if `from_request == from_host == true` then `nil` is returned which means `setup_insights() == false`, doesn't it?

File: app/controllers/registration_command_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+class RegistrationCommandController < ApplicationController
```
Comment (ofedoren) [inline app/controllers/registration_command_controller.rb]: Maybe I've missed discussion, so please excuse me. Why renaming is needed? Even if so, why not `GlobalRegistration` or even `HostRegistrationCommand`? Not a blocker tho, just thoughts so you don't have to rename it in Katello :D

File: app/registries/foreman/access_permissions.rb [hunk 6]
```diff
@@ -305,10 +305,11 @@
                                     :puppetclasses => pc_ajax_actions,
                                     :subnets => subnets_ajax_actions,
                                     :interfaces => [:new, :random_name],
-                                    :registration => [:new, :create],
+                                    :registration_command => [:new, :create],
```
Comment (ofedoren) [inline app/registries/foreman/access_permissions.rb]: From my experience it's not worth to change the name of the permission since it will require a migration to rename or delete the old one. Although I might be wrong, it's better to check in rails console.

**Review by ofedoren — COMMENTED** [review]
Checked with hammer (without Katello part) and it seems to be working, although I have few questions/suggestion in terms of refactoring, but maybe I'm just missing something.

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 2 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: > but still I question why that method exists...

Well I didn't want to have too much logic in the `command` method, so I separated it to the `registration_args`.

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 3 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: Because I want to be sure that `registration_params['setup_...]` will be always `String`, and not `Boolean`.

In typical scenario, `registration_params` will be always `String`, but when you write tests or can the methods from somewhere else you can accidentally set it to `Boolean` , which cause this behavior:
```ruby
registration_params['setup_insights'] = false
return if registration_params['setup_insights'].blank? #=> true
# unreachable code, so it's a bug

registration_params['setup_insights'] = 'false'
return if registration_params['setup_insights'].blank? #=> false
# tada! now you can run logic for setting value of parameter
```
But I agree that the could could be simply:
```
return unless registration_params['setup_insights'].to_s
```

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 4 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: Yeah, the logic is confusing, and trust me, you are not the first one who complains (got many BZs from support already).
This will be completely removed with simpler behavior, see:

> As a user I want to be able to pick one of the three options for flags like REX, insights - force to yes, force to know, use the default value. The command should reflect that and either enforce the parameter value or avoid it entirely. https://bugzilla.redhat.com/show_bug.cgi?id=1919958

File: app/controllers/registration_command_controller.rb (hunk 5 above)
Comment (stejskalleos) [reply app/controllers/registration_command_controller.rb]: > Why renaming is needed? 

It's not _needed_, I just wanted to have it same way as in `api` scope:
`RegistrationController` - for the host registration itself
`RegistrationCommand` - for the command stuff only

With this one would expect that `UI` controllers would be same, right?
But no strong opinions here, just wanted to have it same in both scopes. 

> Why renaming is needed? Even if so, why not GlobalRegistration or even HostRegistrationCommand?

Well, `RegistrationController` is shorter than `GlobalRegistrationController` or `HostRegistrationCommandController`,
plus we still don't have official branding name for this feature, so `Registration` was nice compromise.

File: app/registries/foreman/access_permissions.rb (hunk 6 above)
Comment (stejskalleos) [reply app/registries/foreman/access_permissions.rb]: but this is controller name right? The permission is `:create_hosts`, see:
```ruby
    map.permission :create_hosts,  {:hosts => [:new, :create, :clone].push(*ajax_actions),
                                    :compute_resources_vms => [:import],
                                    :compute_resources => cr_ajax_actions,
                                    :puppetclasses => pc_ajax_actions,
                                    :subnets => subnets_ajax_actions,
                                    :interfaces => [:new, :random_name],
                                    :registration_command => [:new, :create],
                                     :"api/v2/hosts" => [:create],
                                     :"api/v2/interfaces" => [:create],
                                     :"api/v2/tasks" => [:index],
                                     :"api/v2/registration_command" => [:create],
                                     :"api/v2/registration" => [:global, :host]
```

**Comment by ezr-ondrej** [comment]
[test katello]

**Comment by stejskalleos** [comment]
Katello tests will always fail, because of https://github.com/Katello/katello/pull/9108

File: app/controllers/api/v2/registration_command_controller.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+module Api
+  module V2
+    class RegistrationCommandController < V2::BaseController
```
Comment (ezr-ondrej) [inline app/controllers/api/v2/registration_command_controller.rb]: **Just an idea**, I do not insist on this :)
I know `RegistrationCommand` is not a resource, but wouldn't it be better to treat it as one and thus name the controller in **plural** form `RegistrationCommandsController` ?
I'm just worried Rails could do unexpected things somewhere if we don't follow that pattern.
Let me know what you think :)

File: app/views/hosts/index.html.erb [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <%
-  register_btn = display_link_if_authorized(_("Register Host"), { :controller => 'registration', :action => :new }, :class => "btn btn-default")
+  register_btn = display_link_if_authorized(_("Register Host"), { :controller => 'registration_command', :action => :new }, :class => "btn btn-default")
```
Comment (ezr-ondrej) [inline app/views/hosts/index.html.erb]:
```suggestion
  register_btn = display_link_if_authorized(_("Register Host"), hash_for_new_registration_command_path, :class => "btn btn-default")
```

File: app/views/hosts/welcome.html.erb [hunk 9]
```diff
@@ -8,6 +8,6 @@
   <p><%= link_to _('Learn more about this in the documentation.'), documentation_url("4.4Provisioning")%></p>
   <div class="blank-slate-pf-main-action">
     <%= new_link(_('Create Host'), {}, { :class => "btn-lg" }) %>
-    <%= display_link_if_authorized(_("Register Host"), { :controller => 'registration', :action => :new }, :class => "btn btn-default btn-lg") %>
+    <%= display_link_if_authorized(_("Register Host"), { :controller => 'registration_command', :action => :new }, :class => "btn btn-default btn-lg") %>
```
Comment (ezr-ondrej) [inline app/views/hosts/welcome.html.erb]:
```suggestion
    <%= display_link_if_authorized(_("Register Host"), hash_for_new_registration_command_path, :class => "btn btn-default btn-lg") %>
```

File: app/views/registration_command/_form.erb [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <% title _('Register Host') %>
-<%= form_tag({controller: 'registration', action: 'create' }, method: :post, class: 'form-horizontal well') do %>
+<%= form_tag({controller: 'registration_command', action: 'create' }, method: :post, class: 'form-horizontal well') do %>
```
Comment (ezr-ondrej) [inline app/views/registration_command/_form.erb]:
```suggestion
<%= form_tag(registration_commands_path, method: :post, class: 'form-horizontal well') do %>
```
Side note: if we would have the RegistrationCommand as a resource, this would be easier.

File: app/controllers/concerns/foreman/controller/registration_command.rb [hunk 11]
```diff
@@ -0,0 +1,66 @@
+module Foreman::Controller::RegistrationCommand
```
Comment (ezr-ondrej) [inline app/controllers/concerns/foreman/controller/registration_command.rb]: Could this be a separate class, that would get the arguments and build the command?
It just seems as bit too much responsibility for the controller.

Btw such class could most definitelly include `ActiveModel` and act as a resource class (for forms and controllers).

**Review by ezr-ondrej — COMMENTED** [review]
Left some ideas for IMHO better code readability, none of them are hard requirements, let me know what you think, I'd be ok even with followup PR, apart of the conroller rename (to a plural form), that if we decide to do, we should do now as we are already renaming it, so we don't rename twice. I'll test once you ack/nack the ideas :)

File: app/controllers/concerns/foreman/controller/registration_command.rb [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  def host_config_params
+    organization = Organization.authorized(:view_organizations).find(registration_params['organization_id']) if registration_params['organization_id'].present?
+    location = Location.authorized(:view_locations).find(registration_params['location_id']) if registration_params['location_id'].present?
+    host_group = Hostgroup.authorized(:view_hostgroups).find(registration_params['hostgroup_id']) if registration_params["hostgroup_id"].present?
+    operatingsystem = Operatingsystem.authorized(:view_operatingsystems).find(registration_params['operatingsystem_id']) if registration_params["operatingsystem_id"].present?
+
+    Host.new(organization: organization, location: location, hostgroup: host_group, operatingsystem: operatingsystem).params
+  end
+
+  def setup_insights
```
Comment (ezr-ondrej) [inline app/controllers/concerns/foreman/controller/registration_command.rb]: is this even something that should be here, shouldn't it be an extension of this controller/resource from https://github.com/theforeman/foreman_rh_cloud ?

File: app/views/registration_command/_form.erb (hunk 10 above)
Comment (stejskalleos) [reply app/views/registration_command/_form.erb]: In follow up PR form is going to be replaced with React component anyway, see https://projects.theforeman.org/issues/31807

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 12 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: Yes we want to have it in Foreman core since some users may want to use insights without RH cloud plugin.

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 11 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: Created redmine: https://projects.theforeman.org/issues/31849

**Comment by stejskalleos** [comment]
@ezr-ondrej @ofedoren  rebased & updated
Renamed names to plural form + updated the routes.
For moving logic from controller to separate class: https://projects.theforeman.org/issues/31849

File: app/views/hosts/welcome.html.erb [hunk 13]
```diff
@@ -8,6 +8,6 @@
   <p><%= link_to _('Learn more about this in the documentation.'), documentation_url("4.4Provisioning")%></p>
   <div class="blank-slate-pf-main-action">
     <%= new_link(_('Create Host'), {}, { :class => "btn-lg" }) %>
-    <%= display_link_if_authorized(_("Register Host"), { :controller => 'registration', :action => :new }, :class => "btn btn-default btn-lg") %>
+    <%= display_link_if_authorized(_("Register Host"), hash_for_register_hosts_path, :class => "btn btn-default btn-lg") %>
```
Comment (ezr-ondrej) [inline app/views/hosts/welcome.html.erb:11]: I'm sorry I've confused you, but I'm not sure this will work without named routes, so this will probably need to be part of https://projects.theforeman.org/issues/31849 IMO.

**Review by ezr-ondrej — COMMENTED** [review]
Apart of the `hash_for*_path` (that doesn't work if I'm not mistaken) this looks good to me.
I need to test it, but code looks good :+1:

File: app/views/hosts/welcome.html.erb (hunk 13 above)
Comment (stejskalleos) [reply app/views/hosts/welcome.html.erb:11]: Well it works on my setup ... can you give me the error you have?

Comment (ezr-ondrej) [reply app/views/hosts/welcome.html.erb:11]: nope, I can't because it was just a hunch, Rails does not generate named routes for `get` automatically, but only if you use `as:` option, but if it works, then great and don't mind me! :100:

File: app/controllers/concerns/foreman/controller/registration_command.rb (hunk 12 above)
Comment (ezr-ondrej) [reply app/controllers/concerns/foreman/controller/registration_command.rb]: okey :shrug:

**Review by ezr-ondrej — APPROVED** [review]
Works for me @stejskalleos.
Thanks to @ofedoren for doing most of the review :)
